lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+        }
+        ($contigobj, $contigID_hash) = $self->_get_contigs($contigref);
+    }
+    my ($oldfunchash, $oldtype, $types_ref);
+    my %types = ();
+    ($oldfunchash, $oldtype, $types_ref,
+        $inputgenome) = $self->_get_oldfunchash_oldtype_types($inputgenome);
+    %types = %{$types_ref};
+    return ($inputgenome, $contigobj, $contigID_hash, $oldfunchash, $oldtype, \%types);
+}
+
+
+# When the input_obj_ref is referencing to an assembly/contigset object
+sub _create_inputgenome_from_assembly {
+    my ($self, $inputgenome, $input_obj_ref) = @_;
+
+    $inputgenome->{assembly_ref} = $input_obj_ref;  # TOBe confirmed
+    my ($contigobj, $contigID_hash) = $self->_get_contigs($input_obj_ref);
+    return ($inputgenome, $contigobj, $contigID_hash);
+}
+
+sub _util_version {
+	my ($self) = @_;
+	my $VERSION = '0.1.6';
+    return $VERSION;
+}
+
+## Checking on non_coding_features for ones that have undefined type field
+sub _check_NC_features {
+    my ($self, $genome) = @_;
+
+    my $ncoding_features = $genome->{non_coding_features};
+    my $cnt = 0;
+    for my $ncoding_ftr (@{$ncoding_features}) {
+        if(exists($ncoding_ftr->{type})) {
+            $cnt++;
+            #print "Non-conding feature type value= $ncoding_ftr->{type}\n";
+        }
+    }
+    if ($cnt == scalar @{$ncoding_features}) {
+        print "***INFO***: All $cnt non-coding features have defined type**********\n";
+    }
+}
+
+
+## Mapping the contigIDs back to their original (long) names
+sub _remap_contigIDs {
+    my ($self, $contigID_hash, $genome) = @_;
+
+    my $contigs = $genome->{contigs};
+    foreach my $ctg_obj (@{$contigs}) {
+        my $cid = $ctg_obj->{id};
+        $ctg_obj->{id} = $contigID_hash->{$cid};
+        $ctg_obj->{name} = $contigID_hash->{$cid};
+    }
+    return $genome;
+}
+## end helper subs
+#
+
+## begin building the sequence of functions refactored from annotate_process in the RASTImpl
+#
+
+#
+## According to inputs in $parameters and a pre-created default genome object, $inputgenome
+## Returns:
+##  ($rast_details = {
+##      parameters => $parameters,
+##      oldfunchash => $oldfunchash,
+##      oldtype => $oldtype,
+##      types => \%types,
+##      contigobj => $contigobj,
+##      fasta_file => $fasta_file,
+##      gff_file => $gff_file
+##  }, $inputgenomes);
+#
+sub _set_parameters_by_input {
+    my ($self, $parameters, $inputgenome) = @_;
+
+    #print "_set_parameters_by_input's input parameters:\n". Dumper($parameters). "\n";
+    my $contigobj;
+    my $contigID_hash = {};
+    my $oldfunchash = {};
+    my $oldtype     = {};
+    my %types = ();
+    my %rast_details = ();
+
+    # 1. getting the fasta & gff files from $input_obj_ref according to its type
+    my $input_obj_ref = $parameters->{object_ref};
+    return ((), {}) unless ($self->_validate_KB_objref($input_obj_ref));
+    my $input_obj_info = $self->_fetch_object_info($input_obj_ref);
+    return ((), {}) unless defined($input_obj_info);
```

Comment:
Look at what you're returning when the function is successful vs when it returns early. They should be of the same type.

Reply:
That's the purpose: if the type is invalid or if there is no obj info found, return the same type.